### PR TITLE
ROS 2: add multiple topics

### PR DESCRIPTION
**Pull request by dmbuck32**
## Add support for bridging multiple ROS2 topics in a single client

This pull request adds support for bridging multiple MQTT <-> ROS topics in a single client instance. 

This follows the suggestion from this ROS discourse post:
- https://discourse.ros.org/t/use-ros-message-in-parameter-interface/31269/4

### CODE REVIEW

File: mqtt_client/src/MqttClient.ros2.cpp [hunk 1]
```diff
@@ -189,33 +189,43 @@ void MqttClient::loadParameters() {
...
+    declare_parameter(fmt::format("bridge.ros2mqtt.{}.advanced.ros.queue_size", topic_name), rclcpp::ParameterType::PARAMETER_INTEGER, param_desc);
+    param_desc.description = "MQTT QoS value";
+    declare_parameter(fmt::format("bridge.ros2mqtt.{}.advanced.mqtt.qos", topic_name), rclcpp::ParameterType::PARAMETER_INTEGER, param_desc);
+    param_desc.description = "whether to retain MQTT message";
+    declare_parameter(fmt::format("bridge.ros2mqtt.{}.advanced.mqtt.retained", topic_name), rclcpp::ParameterType::PARAMETER_BOOL, param_desc);
+  }
+
+  param_desc.description = "The list of topics to bridge from MQTT to ROS";
+  auto mqtt2ros_topic_names = declare_parameter<std::vector<std::string>>("bridge.mqtt2ros.topics", std::vector<std::string>(), param_desc);
+  for (auto topic_name : mqtt2ros_topic_names)
```
Comment (lreiher) [inline mqtt_client/src/MqttClient.ros2.cpp:214]:
```suggestion
  const auto mqtt2ros_topic_names = declare_parameter<std::vector<std::string>>("bridge.mqtt2ros.topics", std::vector<std::string>(), param_desc);
  for (const auto& topic_name : mqtt2ros_topic_names)
```

File: mqtt_client/src/MqttClient.ros2.cpp [hunk 2]
```diff
@@ -267,102 +277,109 @@ void MqttClient::loadParameters() {
...
-        get_logger(),
-        "Timestamp will not be injected into primitive messages on ROS "
-        "topic '%s'",
-        ros_topic.c_str());
-      ros2mqtt.stamped = false;
-    }
+  for (const auto& topic_name : ros2mqtt_topic_names) {
+    rclcpp::Parameter ros_topic_param, mqtt_topic_param;
+    if (get_parameter(fmt::format("bridge.ros2mqtt.{}.ros_topic", topic_name), ros_topic_param) &&
+        get_parameter(fmt::format("bridge.ros2mqtt.{}.mqtt_topic", topic_name), mqtt_topic_param)) {
```
Comment (lreiher) [inline mqtt_client/src/MqttClient.ros2.cpp:283]:
```suggestion
  for (const auto& topic_name : ros2mqtt_topic_names) {
  
    rclcpp::Parameter ros_topic_param, mqtt_topic_param;
    if (get_parameter(fmt::format("bridge.ros2mqtt.{}.ros_topic", topic_name), ros_topic_param) &&
        get_parameter(fmt::format("bridge.ros2mqtt.{}.mqtt_topic", topic_name), mqtt_topic_param)) {
```

**Review by lreiher — CHANGES_REQUESTED** [review]
Thank you very much for your PR!

I have made some minor suggestions and would be very happy to merge afterwards! :)

Will also try to get to your other PR as soon as possible.

**Comment by lreiher** [comment]
Could you please also modify the GitHub actions `.github/workflows` such that the checks also run on PRs? Apparently we have set it up incorrectly, as no checks are currently being tested in this PR.

Simply change the `on` key to this:
```yml
on: [push, pull_request]
```

File: README.md [hunk 3]
```diff
@@ -285,6 +285,38 @@ bridge:
...
+##### ROS2
+
+```yaml
+bridge:
+  ros2mqtt:            # Object specifying which ROS topics to map to which MQTT topics
+    topics:              # Array specifying which ROS topics to bridge
+      - topic_name         # The topic that should be bridged, corresponds to the sub-object in the YAML
+    topic_name:
+      ros_topic:           # ROS topic whose messages are transformed to MQTT messages
+      mqtt_topic:          # MQTT topic on which the corresponding ROS messages are sent to the broker
```
[the remark this answers is not in the document]
Comment (dmbuck32) [reply README.md:297]: `topic_name` is a placeholder for the name of the topic being bridged between MQTT & ROS.

The idea being that this `topic_name` identifies the bridge between the ROS & MQTT.

In the ping-pong example:

```yaml
ros2mqtt:            
    topics:              
      - ping
      - ping_primative
    ping:
      ros_topic: /ping/ros
      mqtt_topic: pingpong/ros
    ping_primative:
      ros_topic: /ping/primitive
      mqtt_topic: pingpong/primitive
```

Are you suggesting that we do the following?

```yaml
ros2mqtt:            
    topics:              
      - /ping/ros
      - pingpong/ros
    /ping/ros:
      ros_topic: /ping/ros
      mqtt_topic: pingpong/ros
    pingpong/ros:
      ros_topic: /ping/primitive
      mqtt_topic: pingpong/primitive
```
If so, is `/ping/ros` a valid YAML name?

Alternatively, should the documentation be updated to indicate that this is a placeholder?

```yaml
bridge:
  ros2mqtt:            # Object specifying which ROS topics to map to which MQTT topics
    topics:              # Array specifying which ROS topics to bridge
      - {{ topic_name }}         # The topic that should be bridged, corresponds to the sub-object in the YAML
    {{ topic_name }}:
      ros_topic:           # ROS topic whose messages are transformed to MQTT messages
      mqtt_topic:          # MQTT topic on which the corresponding ROS messages are sent to the broker
```

[the remark this answers is not in the document]
Comment (lreiher) [reply README.md:297]: I was suggesting the following. Note that we are only specifying the ros topic information once for each topic this way.

```yml
ros2mqtt:            
    ros_topics:              
      - /ping/ros
      - pingpong/ros
    /ping/ros:
      mqtt_topic: pingpong/ros
    pingpong/ros:
      mqtt_topic: pingpong/primitive
```

I am not 100% sure that this is valid yamland correctly parsed by ROS, but a standard online yaml parser I have just quickly tested didn't complain.

[the remark this answers is not in the document]
Comment (dmbuck32) [reply README.md:297]: Ok, I'll have a look at it this weekend and see if I can get this to work.

[the remark this answers is not in the document]
Comment (lreiher) [reply README.md:297]: Great, thanks a lot! After this PR is merged as well, I will go ahead and make another release.

[the remark this answers is not in the document]
Comment (dmbuck32) [reply README.md]: I was able to get this working, apparently it is possible to have a forward slash in a ROS parameter path.

**Review by lreiher — APPROVED** [review]
Great, thanks a lot!

**Pull request merged**